Userland/Libraries/LibGfx/TGALoader.cpp

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright (c) 2022, Tom Needham <06needhamt@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/String.h>
+#include <LibGfx/TGALoader.h>
+
+namespace Gfx {
+
+enum TGADataType : u8 {
+    None = 0,
+    UncompressedColorMapped = 1,
+    UncompressedRGB = 2,
+    UncompressedBlackAndWhite = 3,
+    RunLengthEncodedColorMapped = 9,
+    RunLengthEncodedRGB = 10,
+    CompressedBlackAndWhite = 11,
+    CompressedColorMapped = 32,
+    CompressedColorMappedFourPass = 33
+};
+
+struct TGAHeader {
+    u8 id_length;
+    u8 color_map_type;
+    enum TGADataType data_type_code;
+    i16 color_map_origin;
+    i16 color_map_length;
+    u8 color_map_depth;
+    i16 x_origin;
+    i16 y_origin;
+    u16 width;
+    u16 height;
+    u8 bits_per_pixel;
+    u8 image_descriptor;
+} __attribute__((packed));
+
+static_assert(sizeof(TGAHeader) == 18);
+
+union TGAPixel {
+    struct TGAColor {
+        u8 blue;
+        u8 green;
+        u8 red;
+        u8 alpha;
+    } components;
+
+    u32 data;
+} __attribute__((packed));
+
+static_assert(sizeof(TGAPixel) == 4);
+
+class TGAReader {
+public:
+    TGAReader(Span<u8 const> data)
+        : m_data(move(data))
+    {
+    }
+
+    ALWAYS_INLINE u8 read_u8()
+    {
+        u8 value = m_data[m_index];
+        m_index++;
+        return value;
+    }
+
+    ALWAYS_INLINE i8 read_i8()
+    {
+        return static_cast<i8>(read_u8());
+    }
+
+    ALWAYS_INLINE u16 read_u16()
+    {
+        return read_u8() | read_u8() << 8;
+    }
+
+    ALWAYS_INLINE i16 read_i16()
+    {
+        return read_i8() | read_i8() << 8;
+    }
+
+    ALWAYS_INLINE u32 read_u32()
+    {
+        return read_u16() | read_u16() << 16;
+    }
+
+    ALWAYS_INLINE i32 read_i32()
+    {
+        return read_i16() | read_i16() << 16;
+    }
+
+    ALWAYS_INLINE TGAPixel read_pixel(u8 bits_per_pixel)
+    {
+        auto pixel = TGAPixel();
+
+        switch (bits_per_pixel) {
+        case 24:
+            pixel.components.blue = read_u8();
+            pixel.components.green = read_u8();
+            pixel.components.red = read_u8();
+            pixel.components.alpha = 0xFF;
+            return pixel;
+
+        case 32:
+            pixel.components.blue = read_u8();
+            pixel.components.green = read_u8();
+            pixel.components.red = read_u8();
+            pixel.components.alpha = read_u8();
+            return pixel;
+
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    size_t index() const
+    {
+        return m_index;
+    }
+
+private:
+    Span<u8 const> m_data;
```

Comment:
Ali pointed out to me that we have a `ReadonlyBytes` in `AK/Forward.h` which is identical to `Span<u8 const>` but a bit nicer. My bad! Otherwise this all looks good to me. :^)